File: integrations/strands-ts/src/hooks.ts
/**
 * hooks.ts — FlowAgentsHooks: the main hook provider for Strands Agents TypeScript SDK.
 *
 * Design: duck-typed against the Strands TS registry/event shapes so the module
 * compiles and tests WITHOUT strands-agents installed. strands-agents is listed
 * as an optional peerDependency in package.json.
 *
 * When strands-agents IS installed, FlowAgentsHooks is a valid HookProvider
 * because it implements registerHooks(registry) which calls:
 *   registry.addCallback(EventClass, callback)
 *
 * Usage (with strands-agents installed):
 *
 *   import { Agent, BeforeInvocationEvent, AfterInvocationEvent,
 *            BeforeToolCallEvent, AfterToolCallEvent } from "@strands-agents/sdk";
 *   import { FlowAgentsHooks } from "@kontourai/flow-agents-strands";
 *
 *   const hooks = new FlowAgentsHooks({ workspace: "." });
 *   const agent = new Agent({ hooks: [hooks] });
 *   // or: agent.addHook(BeforeInvocationEvent, cb);
 *
 * Usage (without strands-agents, e.g. tests):
 *
 *   import { FlowAgentsHooks } from "@kontourai/flow-agents-strands";
 *   const hooks = new FlowAgentsHooks();
 *   // All methods callable; registerHooks() is a no-op without the SDK.
 */

import fs from "node:fs";
import path from "node:path";
import { TelemetrySink } from "./telemetry.js";
import { PolicyGate } from "./policy.js";

// ---------------------------------------------------------------------------
// Duck-typed Strands registry/event interfaces
// These match the Strands TS SDK surface structurally; we do NOT import the
// actual SDK so the module compiles without it being installed.
// ---------------------------------------------------------------------------

/** Minimal duck-type for a Strands-TS event class constructor. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type EventClass = new (...args: any[]) => unknown;

/** Minimal duck-type for a Strands-TS HookRegistry. */
export interface HookRegistry {
  addCallback(eventClass: EventClass, callback: (event: StrandsEvent) => void): void;
}

/** Minimal duck-type for Strands events we handle. */
export interface StrandsEvent {
  // BeforeToolCallEvent fields
  toolName?: string;
  toolInput?: Record<string, unknown>;
  // TS variant: cancellable via cancel property
  cancel?: string;
  // AfterToolCallEvent
  retry?: boolean;
  result?: unknown;
  // Common optional fields
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Kit flow discovery (Issue #32, Decision Q3: option (a))
//
// activateStrandsLocal (src/runtime-adapters.ts) writes kit flow files to
// .flow-agents/runtime/strands/flows/<kit-id>/<asset-id>.flow.json.
// steeringContext() reads those files and surfaces their id + description
// so the agent is aware of available workflow guidance without the hooks
// needing to know anything about the catalog layout.
// ---------------------------------------------------------------------------

interface KitFlowEntry {
  kitId: string;
  assetId: string;
  description: string;
}

function findRepoRoot(start: string): string {
  let current = path.resolve(start);
  for (let i = 0; i < 40; i++) {
    if (
      fs.existsSync(path.join(current, ".git")) ||
      fs.existsSync(path.join(current, "AGENTS.md"))
    ) {
      return current;
    }
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return path.resolve(start);
}

function readKitFlows(flowAgentsDir: string): KitFlowEntry[] {
  const flowsDir = path.join(flowAgentsDir, "runtime", "strands", "flows");
  if (!fs.existsSync(flowsDir)) return [];
  const results: KitFlowEntry[] = [];

  function walkDir(dir: string): void {
    for (const name of fs.readdirSync(dir).sort()) {
      const full = path.join(dir, name);
      const stat = fs.statSync(full);
      if (stat.isDirectory()) {
        walkDir(full);
      } else if (name.endsWith(".flow.json")) {
        try {
          const payload = JSON.parse(fs.readFileSync(full, "utf8")) as Record<string, unknown>;
          const stemName = name.replace(/\.flow\.json$/, "");
          const assetId = typeof payload.id === "string" ? payload.id : stemName;
          const description = typeof payload.description === "string" ? payload.description : "";
          // kit_id is the directory component between flows/ and the file
          const rel = path.relative(flowsDir, full);
          const relParts = rel.split(path.sep);
          const kitId = relParts.length >= 2 ? relParts[0] : "";
          results.push({ kitId, assetId, description });
        } catch {
          // Malformed file — skip silently (fail-open)
        }
      }
    }
  }

  walkDir(flowsDir);
  return results;
}

// ---------------------------------------------------------------------------
// Usage extraction — map a Strands model-call event onto the documented
// Anthropic usage object, defensively across SDK shapes.
// ---------------------------------------------------------------------------

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return value && typeof value === "object" ? (value as Record<string, unknown>) : undefined;
}

function numField(obj: Record<string, unknown> | undefined, ...keys: string[]): number {
  if (!obj) return 0;
  for (const key of keys) {
    const v = obj[key];
    if (typeof v === "number" && Number.isFinite(v)) return v;
  }
  return 0;
}

function strField(obj: Record<string, unknown> | undefined, ...keys: string[]): string | undefined {
  if (!obj) return undefined;
  for (const key of keys) {
    const v = obj[key];
    if (typeof v === "string" && v) return v;
  }
  return undefined;
}

export function extractModelUsage(
  event: StrandsEvent
): { model: string; input: number; output: number; cacheCreation: number; cacheRead: number } | null {
  // Find the usage object wherever the event surfaces it.
  const containers = [
    event,
    asRecord(event.usage),
    asRecord(event.response),
    asRecord(event.result),
    asRecord(event.message),
    asRecord(event.output),
    asRecord(event.modelResponse),
  ];
  let usage: Record<string, unknown> | undefined;
  let modelCarrier: Record<string, unknown> | undefined;
  for (const container of containers) {
    const c = asRecord(container);
    if (!c) continue;
    const candidate = asRecord(c.usage) ?? (("input_tokens" in c || "inputTokens" in c) ? c : undefined);
    if (candidate && !usage) usage = candidate;
    if (!modelCarrier && (typeof c.model === "string" || typeof c.modelId === "string")) modelCarrier = c;
  }
  if (!usage) return null;

  const input = numField(usage, "input_tokens", "inputTokens");
  const output = numField(usage, "output_tokens", "outputTokens");
  const cacheCreation = numField(usage, "cache_creation_input_tokens", "cacheCreationInputTokens");
  const cacheRead = numField(usage, "cache_read_input_tokens", "cacheReadInputTokens");
  if (input === 0 && output === 0 && cacheCreation === 0 && cacheRead === 0) return null;

  const model =
    strField(modelCarrier, "model", "modelId") ??
    strField(usage, "model") ??
    "unknown";
  return { model, input, output, cacheCreation, cacheRead };
}

function buildKitFlowsHint(flows: KitFlowEntry[]): string {
  if (flows.length === 0) return "";
  const lines = ["KIT FLOWS: the following kit flows are activated for this workspace:"];
  for (const flow of flows) {
    const desc = flow.description ? ` — ${flow.description.slice(0, 120)}` : "";
    lines.push(`  • ${flow.assetId}${desc}`);
  }
  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// FlowAgentsHooks options
// ---------------------------------------------------------------------------

export interface FlowAgentsHooksOptions {
  /** JSONL telemetry sink path or directory. Default: <workspace>/.telemetry/full.jsonl */
  sinkPath?: string;
  /** Root of the workspace (reads .flow-agents/ and .telemetry/). Default: process.cwd() */
  workspace?: string;
  /** Agent identifier embedded in telemetry. Default: "strands-agent" */
  agentName?: string;
  /** Runtime label in telemetry. Default: "strands-ts" */
  runtime?: string;
  /**
   * Root of the @kontourai/flow-agents package for native engine import.
   * Defaults to auto-discovery (see policy.ts).
   */
  engineRoot?: string;
}

// ---------------------------------------------------------------------------
// FlowAgentsHooks
// ---------------------------------------------------------------------------

export class FlowAgentsHooks {
  private readonly sink: TelemetrySink;
  private readonly policyGate: PolicyGate;
  private readonly _workspace: string;
  private _sessionStartMs: number | null = null;
  // Per-model token accumulator, summed across model-call events for the session.
  private _usageByModel = new Map<
    string,
    { input: number; output: number; cacheCreation: number; cacheRead: number }
  >();

  constructor(options: FlowAgentsHooksOptions = {}) {
    this._workspace = findRepoRoot(options.workspace ?? process.cwd());

    this.sink = new TelemetrySink({
      sinkPath: options.sinkPath,
      workspace: options.workspace,
      agentName: options.agentName ?? "strands-agent",
      runtime: options.runtime ?? "strands-ts",
    });

    this.policyGate = new PolicyGate({
      engineRoot: options.engineRoot,
    });
  }

  // --------------------------------------------------------------------------
  // Steering context — available without strands-agents installed (Issue #32 AC2)
  // --------------------------------------------------------------------------

  /**
   * Return workflow-steering context text for the current workspace.
   *
   * Includes activated kit flows discovered from the strands-local runtime
   * path (.flow-agents/runtime/strands/flows/) written by
   * `flow-agents kit activate --adapter strands-local`.
   *
   * Callers should prepend this to the Agent's system prompt:
   *
   *   const hooks = new FlowAgentsHooks({ workspace: "." });
   *   const agent = new Agent({ systemPrompt: basePrompt + hooks.steeringContext() });
   */
  steeringContext(): string {
    const flowAgentsDir = path.join(this._workspace, ".flow-agents");
    const flows = readKitFlows(flowAgentsDir);
    const kitFlowsHint = buildKitFlowsHint(flows);
    if (!kitFlowsHint) return "";
    return "\n\n---\n" + kitFlowsHint + "\n---";
  }

  // --------------------------------------------------------------------------
  // HookProvider protocol — registerHooks(registry)
  // This is the sole method required by the Strands HookProvider protocol.
  // --------------------------------------------------------------------------

  /**
   * Register Flow Agents callbacks with a Strands HookRegistry.
   *
   * Requires strands-agents to be installed. If the SDK is absent,
   * this method throws ImportError with instructions.
   */
  registerHooks(registry: HookRegistry): void {
    // Lazily import Strands event classes. If the SDK is not installed,
    // a helpful error is thrown.
    let BeforeInvocationEvent: EventClass;
    let AfterInvocationEvent: EventClass;
    let BeforeToolCallEvent: EventClass;
    let AfterToolCallEvent: EventClass;

    try {
      // Dynamic import is used so the module compiles without the SDK installed.
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const sdk = require("strands-agents") as {
        BeforeInvocationEvent: EventClass;
        AfterInvocationEvent: EventClass;
        BeforeToolCallEvent: EventClass;
        AfterToolCallEvent: EventClass;
      };
      BeforeInvocationEvent = sdk.BeforeInvocationEvent;
      AfterInvocationEvent = sdk.AfterInvocationEvent;
      BeforeToolCallEvent = sdk.BeforeToolCallEvent;
      AfterToolCallEvent = sdk.AfterToolCallEvent;
    } catch (err) {
      throw new Error(
        "strands-agents is required to register hooks. " +
          "Install it with: npm install @strands-agents/sdk\n" +
          `Original error: ${err instanceof Error ? err.message : String(err)}`
      );
    }

    registry.addCallback(BeforeInvocationEvent, (event) => this.onBeforeInvocation(event));
    registry.addCallback(AfterInvocationEvent, (event) => this.onAfterInvocation(event));
    registry.addCallback(BeforeToolCallEvent, (event) => this.onBeforeToolCall(event));
    registry.addCallback(AfterToolCallEvent, (event) => this.onAfterToolCall(event));

    // AfterModelCallEvent carries per-call token usage (the SDK's documented
    // usage source). Optional — only registered if the installed SDK exposes it,
    // so older SDKs still work (usage is simply not collected there).
    // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-explicit-any
    const AfterModelCallEvent = (require("strands-agents") as any).AfterModelCallEvent as EventClass | undefined;
    if (AfterModelCallEvent) {
      registry.addCallback(AfterModelCallEvent, (event) => this.onAfterModelCall(event));
    }
  }

  // --------------------------------------------------------------------------
  // Callbacks — public for direct wiring in tests / without SDK
  // --------------------------------------------------------------------------

  /** BeforeInvocationEvent → userPromptSubmit / turn.user */
  onBeforeInvocation(_event: StrandsEvent): void {
    if (this._sessionStartMs === null) {
      this._sessionStartMs = Date.now();
    }
    this.sink.emitUserPromptSubmit();
  }

  /** AfterInvocationEvent → emit session.usage (if any) then stop / session.end */
  onAfterInvocation(_event: StrandsEvent): void {
    const durationMs =
      this._sessionStartMs !== null ? Date.now() - this._sessionStartMs : 0;

    if (this._usageByModel.size > 0) {
      const byModel = Array.from(this._usageByModel.entries()).map(([model, t]) => ({
        model,
        inputTokens: t.input,
        outputTokens: t.output,
        cacheCreationInputTokens: t.cacheCreation,
        cacheReadInputTokens: t.cacheRead,
      }));
      const sum = byModel.reduce(
        (acc, m) => ({
          input: acc.input + m.inputTokens,
          output: acc.output + m.outputTokens,
          cacheCreation: acc.cacheCreation + m.cacheCreationInputTokens,
          cacheRead: acc.cacheRead + m.cacheReadInputTokens,
        }),
        { input: 0, output: 0, cacheCreation: 0, cacheRead: 0 }
      );
      this.sink.emitUsage({
        model: byModel.length === 1 ? byModel[0].model : undefined,
        durationS: durationMs / 1000,
        inputTokens: sum.input,
        outputTokens: sum.output,
        cacheCreationInputTokens: sum.cacheCreation,
        cacheReadInputTokens: sum.cacheRead,
        byModel,
      });
      this._usageByModel.clear();
    }

    this.sink.emitSessionEnd(durationMs);
  }

  /**
   * AfterModelCallEvent → accumulate per-model token usage.
   *
   * Reads the documented Anthropic usage object (input_tokens, output_tokens,
   * cache_creation_input_tokens, cache_read_input_tokens) from wherever the
   * Strands event surfaces it. Defensive across SDK shapes — if no usage is
   * found, the call is a no-op (tokens for that turn are simply not counted).
   */
  onAfterModelCall(event: StrandsEvent): void {
    const extracted = extractModelUsage(event);
    if (!extracted) return;
    const { model, input, output, cacheCreation, cacheRead } = extracted;
    const current = this._usageByModel.get(model) ?? {
      input: 0,
      output: 0,
      cacheCreation: 0,
      cacheRead: 0,
    };
    current.input += input;
    current.output += output;
    current.cacheCreation += cacheCreation;
    current.cacheRead += cacheRead;
    this._usageByModel.set(model, current);
  }

  /**
   * BeforeToolCallEvent → preToolUse / tool.invoke + config-protection policy gate.
   *
   * If the policy gate blocks the call, sets event.cancel to the block reason
   * (Strands TS variant: event.cancel = "reason").
   */
  onBeforeToolCall(event: StrandsEvent): void {
    const toolName = (event.toolName as string | undefined) ?? "";
    const toolInput = (event.toolInput as Record<string, unknown> | undefined) ?? {};

    // Emit telemetry first (fail-open: policy check follows)
    this.sink.emitToolInvoke(toolName, toolInput);

    // Policy gate — native engine call (no subprocess)
    const blockReason = this.policyGate.checkToolCall(toolName, toolInput);
    if (blockReason) {
      try {
        event.cancel = blockReason;
      } catch {
        // Some event mock or future SDK change; ignore and continue
      }
    }
  }

  /** AfterToolCallEvent → postToolUse / tool.result */
  onAfterToolCall(event: StrandsEvent): void {
    const toolName = (event.toolName as string | undefined) ?? "";
    const result = event.result;
    this.sink.emitToolResult(toolName, result);
  }

  // --------------------------------------------------------------------------
  // Session start — emit agentSpawn when wiring is complete
  // --------------------------------------------------------------------------

  /** Call once after constructing / wiring to emit the agentSpawn event. */
  emitSessionStart(): void {
    this._sessionStartMs = Date.now();
    this._usageByModel.clear();
    this.sink.emitSessionStart();
  }
}

File: integrations/strands/flow_agents_strands/hooks.py
"""
hooks.py — FlowAgentsHooks: the main HookProvider for AWS Strands Agents.

Design: duck-typed so strands-agents is NOT required at import time.
The class uses TYPE_CHECKING guards and string-based isinstance() avoidance so
the full module tree is importable and unit-testable without the SDK installed.

When strands-agents IS installed, FlowAgentsHooks is a valid HookProvider
because it implements the register_hooks(registry, **kwargs) protocol method.

Usage (with strands installed):

    from strands import Agent
    from flow_agents_strands import FlowAgentsHooks

    hooks = FlowAgentsHooks(workspace=".")
    system_prompt = "You are a helpful agent." + hooks.steering_context()
    agent = Agent(system_prompt=system_prompt, hooks=[hooks])

Usage (without strands, e.g. in tests):

    from flow_agents_strands import FlowAgentsHooks
    hooks = FlowAgentsHooks()
    ctx = hooks.steering_context()   # works without strands
"""

from __future__ import annotations

import time
from typing import Any, Callable, Dict, Optional, TYPE_CHECKING

from .telemetry import TelemetrySink, STRANDS_TO_CANONICAL
from .policy import PolicyGate
from .steering import SteeringContext

if TYPE_CHECKING:
    # These imports only run during static type-checking (mypy/pyright).
    # At runtime the try/except below handles the optional SDK.
    from strands.hooks import (  # type: ignore[import]
        HookRegistry,
        BeforeInvocationEvent,
        AfterInvocationEvent,
        BeforeToolCallEvent,
        AfterToolCallEvent,
    )


class FlowAgentsHooks:
    """
    Flow Agents HookProvider for AWS Strands Agents.

    Implements the strands HookProvider protocol (register_hooks) via duck
    typing.  When strands-agents is not installed the class still fully
    constructs and is usable for telemetry emission and steering context
    loading.

    Args:
        sink_path:   Directory or file path for JSONL telemetry output.
                     Default: <workspace>/.flow-agents/.telemetry/full.jsonl
        workspace:   Root of the workspace to discover .flow-agents/ from.
                     Default: current working directory.
        agent_name:  Agent identifier embedded in telemetry events.
        runtime:     Runtime label embedded in telemetry events.
        policy_gate: Optional PolicyGate instance; defaults to PolicyGate().
    """

    def __init__(
        self,
        sink_path: Optional[str] = None,
        workspace: Optional[str] = None,
        agent_name: str = "strands-agent",
        runtime: str = "strands",
        policy_gate: Optional[PolicyGate] = None,
    ) -> None:
        self._sink = TelemetrySink(
            sink_path=sink_path,
            workspace=workspace,
            agent_name=agent_name,
            runtime=runtime,
        )
        self._policy = policy_gate if policy_gate is not None else PolicyGate()
        self._steering = SteeringContext(workspace=workspace)
        self._session_start_ts: Optional[float] = None
        # Per-model token accumulator, summed across model-call events.
        self._usage_by_model: Dict[str, Dict[str, int]] = {}

    # ------------------------------------------------------------------
    # Public API available WITHOUT strands installed
    # ------------------------------------------------------------------

    def steering_context(self) -> str:
        """
        Return workflow-steering context text for the current workspace.

        Callers should append this to the Agent's system prompt at construction
        time — e.g.:

            system_prompt = base_prompt + hooks.steering_context()

        This is the documented spike approach because Strands'
        BeforeInvocationEvent does not expose a mutable system_prompt.
        See README.md § Limitations.
        """
        text = self._steering.load()
        if text:
            self._sink.emit_steering(text)
        return text

    # ------------------------------------------------------------------
    # HookProvider protocol (register_hooks)
    # ------------------------------------------------------------------

    def register_hooks(self, registry: Any, **kwargs: Any) -> None:
        """
        Register Flow Agents callbacks with a Strands HookRegistry.

        This method is the sole method required by the HookProvider protocol.
        The registry parameter is typed as Any so the module compiles without
        strands-agents installed; at runtime the real HookRegistry is passed.
        """
        # Import lazily — only reachable when strands IS installed.
        try:
            from strands.hooks import (  # type: ignore[import]
                BeforeInvocationEvent,
                AfterInvocationEvent,
                BeforeToolCallEvent,
                AfterToolCallEvent,
                AgentInitializedEvent,
            )
        except ImportError as exc:
            raise ImportError(
                "strands-agents is required to register hooks. "
                "Install it with: pip install flow-agents-strands[strands]"
            ) from exc

        registry.add_callback(AgentInitializedEvent, self._on_agent_initialized)
        registry.add_callback(BeforeInvocationEvent, self._on_before_invocation)
        registry.add_callback(AfterInvocationEvent, self._on_after_invocation)
        registry.add_callback(BeforeToolCallEvent, self._on_before_tool_call)
        registry.add_callback(AfterToolCallEvent, self._on_after_tool_call)

        # Model-call event carries per-call token usage (the SDK's documented
        # usage source). Optional — registered only if the installed SDK exposes
        # it, under whichever name this SDK version uses.
        try:
            import strands.hooks as _sh  # type: ignore[import]

            model_event = (
                getattr(_sh, "AfterModelCallEvent", None)
                or getattr(_sh, "AfterModelInvocationEvent", None)
            )
            if model_event is not None:
                registry.add_callback(model_event, self._on_after_model_call)
        except ImportError:
            pass

    # ------------------------------------------------------------------
    # Private callbacks
    # ------------------------------------------------------------------

    def _on_agent_initialized(self, event: Any) -> None:
        """AgentInitializedEvent → agentSpawn / session.start"""
        self._session_start_ts = time.monotonic()
        self._usage_by_model = {}
        self._sink.emit_session_start()

    def _on_before_invocation(self, event: Any) -> None:
        """BeforeInvocationEvent → userPromptSubmit / turn.user"""
        if self._session_start_ts is None:
            self._session_start_ts = time.monotonic()
        self._sink.emit("userPromptSubmit")

    def _on_after_invocation(self, event: Any) -> None:
        """AfterInvocationEvent → emit session.usage (if any) then stop / session.end"""
        duration_s = 0.0
        if self._session_start_ts is not None:
            duration_s = time.monotonic() - self._session_start_ts

        if self._usage_by_model:
            by_model = []
            totals = {"input": 0, "output": 0, "cache_creation": 0, "cache_read": 0}
            for model, tok in self._usage_by_model.items():
                by_model.append(
                    {
                        "model": model,
                        "input_tokens": tok["input"],
                        "output_tokens": tok["output"],
                        "cache_creation_input_tokens": tok["cache_creation"],
                        "cache_read_input_tokens": tok["cache_read"],
                    }
                )
                for key in totals:
                    totals[key] += tok[key]
            self._sink.emit_usage(
                model=next(iter(self._usage_by_model)) if len(self._usage_by_model) == 1 else None,
                input_tokens=totals["input"],
                output_tokens=totals["output"],
                cache_creation_input_tokens=totals["cache_creation"],
                cache_read_input_tokens=totals["cache_read"],
                duration_s=duration_s,
                by_model=by_model,
            )
            self._usage_by_model = {}

        self._sink.emit_session_end(duration_s=duration_s)

    def _on_after_model_call(self, event: Any) -> None:
        """Model-call event → accumulate per-model token usage.

        Reads the documented Anthropic usage object (input_tokens, output_tokens,
        cache_creation_input_tokens, cache_read_input_tokens) from wherever the
        Strands event surfaces it. Defensive across SDK shapes; no-op if absent.
        """
        extracted = _extract_model_usage(event)
        if extracted is None:
            return
        model = extracted["model"]
        acc = self._usage_by_model.setdefault(
            model, {"input": 0, "output": 0, "cache_creation": 0, "cache_read": 0}
        )
        acc["input"] += extracted["input"]
        acc["output"] += extracted["output"]
        acc["cache_creation"] += extracted["cache_creation"]
        acc["cache_read"] += extracted["cache_read"]

    def _on_before_tool_call(self, event: Any) -> None:
        """
        BeforeToolCallEvent → preToolUse / tool.invoke + policy gate.

        If the policy gate blocks the call, sets event.cancel_tool to the
        block reason (Strands will cancel the tool and return the message
        as the tool result).
        """
        tool_use = getattr(event, "tool_use", {}) or {}
        tool_name = tool_use.get("name", "")
        tool_input = tool_use.get("input", {}) or {}

        # Emit telemetry first (fail-open: policy check follows)
        self._sink.emit_tool_invoke(tool_name=tool_name, tool_input=tool_input)

        # Policy gate
        block_reason = self._policy.check_tool_call(
            tool_name=tool_name,
            tool_input=tool_input,
        )
        if block_reason:
            try:
                event.cancel_tool = block_reason
            except AttributeError:
                # Some event mock or future SDK change; log and continue
                pass

    def _on_after_tool_call(self, event: Any) -> None:
        """AfterToolCallEvent → postToolUse / tool.result"""
        tool_use = getattr(event, "tool_use", {}) or {}
        tool_name = tool_use.get("name", "")
        result = getattr(event, "result", None)
        self._sink.emit_tool_result(tool_name=tool_name, tool_output=result)


# ----------------------------------------------------------------------------
# Usage extraction — map a Strands model-call event onto the documented
# Anthropic usage object, defensively across SDK shapes (object or dict).
# ----------------------------------------------------------------------------


def _attr(obj: Any, *keys: str) -> Any:
    for key in keys:
        if isinstance(obj, dict):
            if key in obj and obj[key] is not None:
                return obj[key]
        else:
            value = getattr(obj, key, None)
            if value is not None:
                return value
    return None


def _num(obj: Any, *keys: str) -> int:
    value = _attr(obj, *keys)
    return value if isinstance(value, (int, float)) else 0


def _extract_model_usage(event: Any) -> Optional[Dict[str, Any]]:
    containers = [
        event,
        _attr(event, "usage"),
        _attr(event, "response"),
        _attr(event, "result"),
        _attr(event, "message"),
        _attr(event, "output"),
        _attr(event, "model_response"),
    ]
    usage = None
    model_carrier = None
    for container in containers:
        if container is None:
            continue
        candidate = _attr(container, "usage")
        if candidate is None and (_attr(container, "input_tokens", "inputTokens") is not None):
            candidate = container
        if candidate is not None and usage is None:
            usage = candidate
        if model_carrier is None and _attr(container, "model", "model_id", "modelId") is not None:
            model_carrier = container
    if usage is None:
        return None

    tokens = {
        "input": _num(usage, "input_tokens", "inputTokens"),
        "output": _num(usage, "output_tokens", "outputTokens"),
        "cache_creation": _num(usage, "cache_creation_input_tokens", "cacheCreationInputTokens"),
        "cache_read": _num(usage, "cache_read_input_tokens", "cacheReadInputTokens"),
    }
    if not any(tokens.values()):
        return None

    model = _attr(model_carrier, "model", "model_id", "modelId") or _attr(usage, "model") or "unknown"
    return {"model": str(model), **tokens}
